pkg/ds/farm.go

```diff
@@ -0,0 +1,358 @@
+package ds
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/square/p2/pkg/util"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/square/p2/pkg/alerting"
+	"github.com/square/p2/pkg/ds/fields"
+	ds_fields "github.com/square/p2/pkg/ds/fields"
+	"github.com/square/p2/pkg/kp"
+	"github.com/square/p2/pkg/kp/dsstore"
+	"github.com/square/p2/pkg/labels"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/scheduler"
+	"github.com/square/p2/pkg/types"
+	klabels "k8s.io/kubernetes/pkg/labels"
+)
+
+// Farm instatiates and deletes daemon sets as needed
+type Farm struct {
+	// constructor arguments
+	kpStore    kp.Store
+	dsStore    dsstore.Store
+	scheduler  scheduler.Scheduler
+	applicator labels.Applicator
+
+	children map[fields.ID]*childDS
+	childMu  sync.Mutex
+
+	logger  logging.Logger
+	alerter alerting.Alerter
+}
+
+type childDS struct {
+	ds        DaemonSet
+	quitCh    chan<- struct{}
+	updatedCh chan<- *ds_fields.DaemonSet
+	deletedCh chan<- *ds_fields.DaemonSet
+	errCh     <-chan error
+}
+
+func NewFarm(
+	kpStore kp.Store,
+	dsStore dsstore.Store,
+	applicator labels.Applicator,
+	logger logging.Logger,
+	alerter alerting.Alerter,
+) *Farm {
+	if alerter == nil {
+		alerter = alerting.NewNop()
+	}
+
+	return &Farm{
+		kpStore:    kpStore,
+		dsStore:    dsStore,
+		scheduler:  scheduler.NewApplicatorScheduler(applicator),
+		applicator: applicator,
+		children:   make(map[fields.ID]*childDS),
+		logger:     logger,
+		alerter:    alerter,
+	}
+}
+
+func (dsf *Farm) Start(quitCh <-chan struct{}) {
+	dsf.cleanupDaemonSetPods(quitCh)
+	dsf.mainLoop(quitCh)
+}
+
+const cleanupInterval = 60 * time.Second
+
+// This function removes all pods with a DSIDLabel where the daemon set id does
+// not exist in the store at every interval specified because it is possible
+// that the farm will unexpectedly crash or someone deletes or modifies a node
+func (dsf *Farm) cleanupDaemonSetPods(quitCh <-chan struct{}) {
+	go func() {
+		timer := time.NewTimer(time.Duration(0))
+
+		for {
+			select {
+			case <-quitCh:
+				return
+			case <-timer.C:
+			}
+			timer.Reset(cleanupInterval)
+
+			allDaemonSets, err := dsf.dsStore.List()
+			dsIDMap := make(map[string]ds_fields.DaemonSet)
+			for _, dsFields := range allDaemonSets {
+				dsIDMap[dsFields.ID.String()] = dsFields
+			}
+
+			allPods, err := dsf.applicator.GetMatches(klabels.Everything(), labels.POD)
+			if err != nil {
+				dsf.logger.NoFields().Fatal(err)
```

Comment:
I think we want `Error` instead of `Fatal` here. `Fatal` calls `os.Exit(1)` which will kill the whole farm. Considering that this is inside of a `for {}` loop i'd think you should log the error and then `continue`